File: segmentacao_textura.cpp
#include <opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include <algorithm>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <limits>
#include <numeric>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace fs = std::filesystem;

constexpr int PATCH_SIZE = 512;
constexpr int FILTER_COUNT = 8;
constexpr int SCALE_COUNT = 3;
constexpr int DESCRIPTOR_SIZE = FILTER_COUNT * SCALE_COUNT;

struct Patch {
    int row;
    int column;
    cv::Mat image;
};

struct DescriptorData {
    cv::Mat values;
    std::vector<Patch> patches;
};

std::vector<cv::Mat> buildFilters() {
    std::vector<cv::Mat> filters;
    filters.emplace_back((cv::Mat_<float>(3, 3) << 1, 2, 1, 0, 0, 0, -1, -2, -1));
    filters.emplace_back((cv::Mat_<float>(3, 3) << 1, 0, -1, 2, 0, -2, 1, 0, -1));
    filters.emplace_back((cv::Mat_<float>(3, 3) << 0, 1, 2, -1, 0, 1, -2, -1, 0));
    filters.emplace_back((cv::Mat_<float>(3, 3) << 2, 1, 0, 1, 0, -1, 0, -1, -2));
    filters.emplace_back((cv::Mat_<float>(3, 3) << 0, 1, 0, 1, -4, 1, 0, 1, 0));

    cv::Mat logKernel(7, 7, CV_32F);
    cv::Mat diffusion(7, 7, CV_32F);
    cv::Mat circular = cv::Mat::zeros(7, 7, CV_32F);
    float logSum = 0.0F;
    float diffusionSum = 0.0F;
    for (int y = -3; y <= 3; ++y) {
        for (int x = -3; x <= 3; ++x) {
            const float radius2 = static_cast<float>(x * x + y * y);
            const float logValue = (radius2 - 4.0F) * std::exp(-radius2 / 8.0F);
            const float diffusionValue = std::exp(-radius2 / 4.0F);
            logKernel.at<float>(y + 3, x + 3) = logValue;
            diffusion.at<float>(y + 3, x + 3) = diffusionValue;
            logSum += logValue;
            diffusionSum += diffusionValue;
            if (radius2 >= 4.0F && radius2 <= 9.0F) {
                circular.at<float>(y + 3, x + 3) = 1.0F;
            }
        }
    }
    logKernel -= logSum / 49.0F;
    diffusion -= diffusionSum / 49.0F;
    circular.at<float>(3, 3) = -static_cast<float>(cv::sum(circular)[0]);
    filters.push_back(logKernel);
    filters.push_back(diffusion);
    filters.push_back(circular);
    return filters;
}

std::vector<Patch> makePatches(const cv::Mat& image, int requested) {
    const int columns = std::min(static_cast<int>(std::sqrt(requested *
        static_cast<double>(image.cols) / image.rows)), image.cols / PATCH_SIZE);
    const int rows = std::min(static_cast<int>(std::ceil(requested / static_cast<double>(std::max(columns, 1)))),
                              image.rows / PATCH_SIZE);
    std::vector<Patch> patches;
    for (int row = 0; row < rows && static_cast<int>(patches.size()) < requested; ++row) {
        for (int column = 0; column < columns && static_cast<int>(patches.size()) < requested; ++column) {
            cv::Rect region(column * PATCH_SIZE, row * PATCH_SIZE, PATCH_SIZE, PATCH_SIZE);
            patches.push_back({row * PATCH_SIZE, column * PATCH_SIZE, image(region).clone()});
        }
    }
    return patches;
}

double meanAbsolute(const cv::Mat& image) {
    return cv::mean(cv::abs(image))[0];
}

DescriptorData extractDescriptors(const cv::Mat& image, int patchCount) {
    const auto filters = buildFilters();
    const auto patches = makePatches(image, patchCount);
    cv::Mat descriptors(static_cast<int>(patches.size()), DESCRIPTOR_SIZE, CV_64F, cv::Scalar(0));
    cv::Mat current = image.clone();
    for (int scale = 0; scale < SCALE_COUNT; ++scale) {
        if (scale > 0) {
            cv::Mat next;
            cv::pyrDown(current, next);
            current = next;
        }
        std::vector<cv::Mat> responses;
        for (const auto& filter : filters) {
            cv::Mat response;
            cv::filter2D(current, response, CV_32F, filter, cv::Point(-1, -1), 0, cv::BORDER_REFLECT101);
            responses.push_back(response);
        }
        const int scaledPatch = PATCH_SIZE >> scale;
        for (int patchIndex = 0; patchIndex < static_cast<int>(patches.size()); ++patchIndex) {
            const int y = patches[patchIndex].row >> scale;
            const int x = patches[patchIndex].column >> scale;
            const cv::Rect region(x, y, scaledPatch, scaledPatch);
            for (int filterIndex = 0; filterIndex < FILTER_COUNT; ++filterIndex) {
                descriptors.at<double>(patchIndex, scale * FILTER_COUNT + filterIndex) =
                    meanAbsolute(responses[filterIndex](region));
            }
        }
    }
    return {descriptors, patches};
}

cv::Mat standardize(const cv::Mat& values) {
    cv::Mat standardized = values.clone();
    for (int column = 0; column < values.cols; ++column) {
        cv::Scalar mean, stddev;
        cv::meanStdDev(values.col(column), mean, stddev);
        const double scale = stddev[0] > 1e-12 ? stddev[0] : 1.0;
        for (int row = 0; row < values.rows; ++row) {
            standardized.at<double>(row, column) = (values.at<double>(row, column) - mean[0]) / scale;
        }
    }
    return standardized;
}

void writeNpy(const fs::path& path, const cv::Mat& values) {
    std::ofstream file(path, std::ios::binary);
    const std::string header = "{'descr': '<f8', 'fortran_order': False, 'shape': (" +
        std::to_string(values.rows) + ", " + std::to_string(values.cols) + "), }";
    const int padding = 16 - ((10 + header.size() + 1) % 16);
    const std::string padded = header + std::string(padding, ' ') + "\n";
    file.write("\x93NUMPY", 6);
    file.put(1); file.put(0);
    const uint16_t headerSize = static_cast<uint16_t>(padded.size());
    file.write(reinterpret_cast<const char*>(&headerSize), sizeof(headerSize));
    file.write(padded.data(), static_cast<std::streamsize>(padded.size()));
    for (int row = 0; row < values.rows; ++row) {
        file.write(reinterpret_cast<const char*>(values.ptr<double>(row)),
                   static_cast<std::streamsize>(values.cols * sizeof(double)));
    }
}

void writeCsv(const fs::path& path, const DescriptorData& data, const cv::Mat& labels) {
    static const std::vector<std::string> filters = {
        "horizontal", "vertical", "45_graus", "135_graus", "laplaciano", "log", "difusao", "circular"
    };
    std::ofstream file(path);
    file << "recorte,linha,coluna";
    for (int scale = 1; scale <= SCALE_COUNT; ++scale) {
        for (const auto& filter : filters) file << ",escala" << scale << "_" << filter;
    }
    file << ",grupo\n" << std::fixed << std::setprecision(8);
    for (int row = 0; row < data.values.rows; ++row) {
        file << row << "," << data.patches[row].row << "," << data.patches[row].column;
        for (int column = 0; column < data.values.cols; ++column) file << "," << data.values.at<double>(row, column);
        file << "," << labels.at<int>(row, 0) + 1 << "\n";
    }
}

cv::Scalar colorFor(int label) {
    static const cv::Scalar colors[] = {
        {57, 70, 230}, {138, 123, 29}, {97, 162, 244}, {143, 157, 42}, {144, 117, 87}
    };
    return colors[label % 5];
}

void saveResults(const fs::path& output, const cv::Mat& image, const DescriptorData& data, const cv::Mat& labels) {
    cv::Mat overlay;
    cv::cvtColor(image, overlay, cv::COLOR_GRAY2BGR);
    const int columns = 8;
    const int tileSize = 180;
    const int rows = static_cast<int>(std::ceil(data.patches.size() / static_cast<double>(columns)));
    cv::Mat montage(rows * tileSize, columns * tileSize, CV_8UC3, cv::Scalar(30, 30, 30));
    for (int index = 0; index < static_cast<int>(data.patches.size()); ++index) {
        const auto& patch = data.patches[index];
        const cv::Scalar color = colorFor(labels.at<int>(index, 0));
        cv::rectangle(overlay, cv::Rect(patch.column, patch.row, PATCH_SIZE, PATCH_SIZE), color, 8);
        cv::putText(overlay, std::to_string(labels.at<int>(index, 0) + 1),
                    cv::Point(patch.column + 15, patch.row + 42), cv::FONT_HERSHEY_SIMPLEX, 1.2, color, 3);
        cv::Mat tile;
        cv::resize(patch.image, tile, cv::Size(tileSize, tileSize));
        cv::cvtColor(tile, tile, cv::COLOR_GRAY2BGR);
        cv::putText(tile, "recorte " + std::to_string(index), cv::Point(7, 22), cv::FONT_HERSHEY_SIMPLEX, 0.45, color, 1);
        cv::putText(tile, "grupo " + std::to_string(labels.at<int>(index, 0) + 1), cv::Point(7, 43), cv::FONT_HERSHEY_SIMPLEX, 0.45, color, 1);
        tile.copyTo(montage(cv::Rect((index % columns) * tileSize, (index / columns) * tileSize, tileSize, tileSize)));
    }
    cv::imwrite((output / "categorizacao_imagem.jpg").string(), overlay, {cv::IMWRITE_JPEG_QUALITY, 90});
    cv::imwrite((output / "categorizacao_recortes.jpg").string(), montage, {cv::IMWRITE_JPEG_QUALITY, 92});
}

int main(int argc, char** argv) {
    try {
        const fs::path imagePath = argc > 1 ? argv[1] : "image.jpg";
        const fs::path output = argc > 2 ? argv[2] : "resultados_cpp";
        const int patchCount = argc > 3 ? std::stoi(argv[3]) : 32;
        const int groups = argc > 4 ? std::stoi(argv[4]) : 4;
        fs::create_directories(output);

        cv::Mat input = cv::imread(imagePath.string(), cv::IMREAD_GRAYSCALE);
        if (input.empty()) throw std::runtime_error("Nao foi possivel abrir " + imagePath.string());
        if (input.cols < PATCH_SIZE || input.rows < PATCH_SIZE) throw std::runtime_error("A imagem deve ter pelo menos 512x512 pixels.");

        const DescriptorData data = extractDescriptors(input, patchCount);
        if (data.values.rows < groups) throw std::runtime_error("Numero de recortes menor que o numero de grupos.");
        const cv::Mat standardizedDouble = standardize(data.values);
        cv::Mat standardized;
        standardizedDouble.convertTo(standardized, CV_32F);
        cv::Mat labels, centers;
        const double compactness = cv::kmeans(standardized, groups, labels,
            cv::TermCriteria(cv::TermCriteria::EPS + cv::TermCriteria::MAX_ITER, 100, 1e-5),
            20, cv::KMEANS_PP_CENTERS, centers);

        writeCsv(output / "descritores_24d.csv", data, labels);
        writeNpy(output / "descritores_24d.npy", data.values);
        saveResults(output, input, data, labels);
        std::cout << "Imagem: " << input.cols << "x" << input.rows << " pixels, tons de cinza\n"
                  << "Recortes analisados: " << data.values.rows << " de 512x512\n"
                  << "Descritores: " << data.values.rows << "x" << data.values.cols << " (8 filtros x 3 escalas)\n"
                  << "Agrupamento: K-Means, " << groups << " grupos, distancia euclidiana padronizada\n"
                  << "Compactacao final: " << compactness << "\n"
                  << "Resultados: " << fs::absolute(output) << "\n";
    } catch (const std::exception& error) {
        std::cerr << "Erro: " << error.what() << "\n";
        return 1;
    }
    return 0;
}

File: segmentacao_textura.py
"""Segmentacao classica por textura para uma fotografia.

O descritor de cada recorte tem 24 dimensoes: 8 filtros x 3 escalas.
"""

from __future__ import annotations

import argparse
import csv
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
from PIL import Image, ImageDraw, ImageOps
from scipy import ndimage
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler


FILTER_NAMES = (
    "horizontal",
    "vertical",
    "diagonal_45",
    "diagonal_135",
    "laplaciano",
    "log",
    "difusao",
    "circular",
)


def build_filters() -> list[np.ndarray]:
    """Cria oito filtros de textura com suporte 7x7."""
    horizontal = np.array([[1, 2, 1], [0, 0, 0], [-1, -2, -1]], dtype=float)
    vertical = horizontal.T
    diagonal_45 = np.array([[0, 1, 2], [-1, 0, 1], [-2, -1, 0]], dtype=float)
    diagonal_135 = diagonal_45.T
    laplaciano = np.array([[0, 1, 0], [1, -4, 1], [0, 1, 0]], dtype=float)

    coordinates = np.arange(-3, 4)
    xx, yy = np.meshgrid(coordinates, coordinates)
    radius2 = xx**2 + yy**2
    log_filter = (radius2 - 4) * np.exp(-radius2 / 8)
    diffusion = np.exp(-radius2 / 4)
    diffusion[3, 3] -= diffusion.sum()
    circular = ((radius2 >= 4) & (radius2 <= 9)).astype(float)
    circular[3, 3] = -circular.sum()

    return [
        horizontal,
        vertical,
        diagonal_45,
        diagonal_135,
        laplaciano,
        log_filter,
        diffusion,
        circular,
    ]


def load_gray(path: Path) -> np.ndarray:
    image = ImageOps.exif_transpose(Image.open(path)).convert("L")
    return np.asarray(image, dtype=np.float32) / 255.0


def make_patches(image: np.ndarray, size: int, count: int) -> list[tuple[int, int, np.ndarray]]:
    """Seleciona recortes 512x512 em uma grade regular, sem ampliar a foto."""
    height, width = image.shape
    columns = max(1, int(np.sqrt(count * width / height)))
    rows = max(1, int(np.ceil(count / columns)))
    columns = min(columns, width // size)
    rows = min(rows, height // size)
    positions = [(row * size, column * size) for row in range(rows) for column in range(columns)]
    positions = positions[:count]
    return [(y, x, image[y : y + size, x : x + size]) for y, x in positions]


def resize_array(array: np.ndarray, shape: tuple[int, int]) -> np.ndarray:
    zoom = (shape[0] / array.shape[0], shape[1] / array.shape[1])
    return ndimage.zoom(array, zoom, order=1)


def extract_descriptors(
    image: np.ndarray, patches: list[tuple[int, int, np.ndarray]], patch_size: int
) -> np.ndarray:
    filters = build_filters()
    descriptors = np.zeros((len(patches), len(filters) * 3), dtype=float)
    for scale_index, scale in enumerate((1.0, 0.5, 0.25)):
        scaled_shape = (round(image.shape[0] * scale), round(image.shape[1] * scale))
        scaled = ndimage.gaussian_filter(image, sigma=1.0) if scale == 1.0 else resize_array(
            ndimage.gaussian_filter(image, sigma=1.0), scaled_shape
        )
        filtered = [ndimage.convolve(scaled, kernel, mode="reflect") for kernel in filters]
        for patch_index, (y, x, _) in enumerate(patches):
            scaled_y, scaled_x = round(y * scale), round(x * scale)
            scaled_size = max(1, round(patch_size * scale))
            for filter_index, response in enumerate(filtered):
                region = response[scaled_y : scaled_y + scaled_size, scaled_x : scaled_x + scaled_size]
                descriptors[patch_index, scale_index * len(filters) + filter_index] = np.mean(
                    np.abs(region)
                )
    return descriptors


def save_csv(
    path: Path,
    patches: list[tuple[int, int, np.ndarray]],
    descriptors: np.ndarray,
    labels: np.ndarray,
) -> None:
    with path.open("w", newline="", encoding="utf-8") as file:
        writer = csv.writer(file)
        writer.writerow(["recorte", "linha", "coluna", *[f"f{i:02d}" for i in range(24)], "grupo"])
        for index, ((y, x, _), descriptor, label) in enumerate(zip(patches, descriptors, labels)):
            writer.writerow([index, y, x, *[f"{value:.8f}" for value in descriptor], int(label)])


def save_montage(path: Path, patches: list[tuple[int, int, np.ndarray]], labels: np.ndarray) -> None:
    columns = 8
    rows = int(np.ceil(len(patches) / columns))
    figure, axes = plt.subplots(rows, columns, figsize=(16, 2 * rows))
    axes = np.atleast_1d(axes).ravel()
    colors = plt.get_cmap("tab10")
    for index, ((_, _, patch), label) in enumerate(zip(patches, labels)):
        axes[index].imshow(patch, cmap="gray", vmin=0, vmax=1)
        axes[index].set_title(f"recorte {index} | grupo {label + 1}", color=colors(label))
        axes[index].axis("off")
    for axis in axes[len(patches) :]:
        axis.axis("off")
    figure.tight_layout()
    figure.savefig(path, dpi=150)
    plt.close(figure)


def save_overlay(path: Path, image: np.ndarray, patches: list[tuple[int, int, np.ndarray]], labels: np.ndarray) -> None:
    output = Image.fromarray(np.uint8(image * 255), mode="L").convert("RGB")
    draw = ImageDraw.Draw(output)
    colors = [(230, 57, 70), (29, 123, 138), (244, 162, 97), (42, 157, 143), (87, 117, 144)]
    for index, ((y, x, _), label) in enumerate(zip(patches, labels)):
        color = colors[int(label) % len(colors)]
        draw.rectangle((x, y, x + 512, y + 512), outline=color, width=8)
        draw.text((x + 12, y + 12), str(int(label) + 1), fill=color)
    output.save(path, quality=90, optimize=True)


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--image", type=Path, default=Path("image.jpg"))
    parser.add_argument("--output", type=Path, default=Path("resultados"))
    parser.add_argument("--patches", type=int, default=32)
    parser.add_argument("--groups", type=int, default=4)
    args = parser.parse_args()

    args.output.mkdir(exist_ok=True)
    image = load_gray(args.image)
    patches = make_patches(image, size=512, count=args.patches)
    if len(patches) < args.groups:
        raise ValueError("O numero de recortes deve ser maior ou igual ao numero de grupos.")

    descriptors = extract_descriptors(image, patches, patch_size=512)
    standardized = StandardScaler().fit_transform(descriptors)
    model = KMeans(n_clusters=args.groups, n_init=20, random_state=42)
    labels = model.fit_predict(standardized)

    np.save(args.output / "descritores_24d.npy", descriptors)
    save_csv(args.output / "descritores_24d.csv", patches, descriptors, labels)
    save_montage(args.output / "categorizacao_recortes.png", patches, labels)
    save_overlay(args.output / "categorizacao_imagem.jpg", image, patches, labels)
    print(f"Imagem: {image.shape[1]}x{image.shape[0]} pixels, tons de cinza")
    print(f"Recortes analisados: {len(patches)}")
    print(f"Descritores: {descriptors.shape} (8 filtros x 3 escalas)")
    print(f"Grupos: {args.groups}")
    print(f"Resultados salvos em: {args.output.resolve()}")


if __name__ == "__main__":
    main()